### ISSUE TITLE
There is no ODR violation when combining ABI-incompatible versions of a library in the same binary via cppwinrt

### ISSUE THREAD

**Issue by dmachaj** [microsoft/cppwinrt#1176]
This issue was encountered when mixing WinUI2 and WinUI3 in the same binary.  I believe the same problem would apply if different WinAppSDK versions were accidentally mixed (e.g. 1.0 is not ABI compatible with 1.1).

The problem is that illegally mixing these types does not break the build when they are in separate static libraries.  If foo.lib uses WinUI2, and bar.lib use WinUI3, then they will link together without errors and the linker will pick one or the other definitions as the winner.  The ABI is not compatible so this will explode horribly at runtime.  Or even worse they might be close enough that it partially works and confuses everyone when it sometimes explodes and sometimes does not.

Based on my (imperfect) understanding of C++ this is a classic ODR violation.  There is a “#pragma detect mismatch” which can be used to turn this type of silent violation into a build break.  However, that pragma would have to be embedded in the cppwinrt headers because that is what gets used by these libs.  (The mixing will be in a cppwinrt code base, so CX and WRL are not relevant to this discussion).

I created a sample project that demonstrates this issue (available upon request; currently in a private repo).  I began with the WinAppSDK desktop app template in Visual Studio 2022.  I then added a WinUI2 static library as a second project in the same solution.  The versions of everything were then updated to the latest (e.g. WinAppSDK is 1.1.3, cppwinrt is 2.0.220608.4, WinUI is 2.7.2).  I added stub usage of `winrt::Microsoft::UI::Xaml::Controls::Primitives::ColorPickerSlider` to force that type to show up in the final binary.  

I opened the resulting binary in [SizeBench](https://aka.ms/sizebench) and analyzed it.  The `types` breakdown only has one `ColorPickerSlider` type.  That type is sized to the WinUI2 ABI.  The WinUI3 ABI type is not present.  As a result I would expect any usage of the WinUI3 `ColorPickerSlider` to explode at runtime.  (The sizes were determined by also creating separate WinUI2 and WinUI3 projects and analyzing them the same way).

**Comment by kennykerr** [comment]
Perhaps something @Scottj1s would like to look into? I generally have no bandwidth to deal with Xaml issues. Alternatively, we can transfer to the Xaml repo.

**Comment by dmachaj** [comment]
@kennykerr note that this is not XAML specific; that is simply where it was discovered.  The same issue could occur without using XAML at all.  A program that accidentally mixes WinAppSDK 1.0 and 1.1 from different static libraries would hit this problem for any types that have changed (e.g. push notifications, MRTCore).

@Scottj1s what do you think?

**Comment by sylveon** [comment]
I don't believe that just WASDK version mixes would have caused it, as the vtable layout & interface GUIDs for existing member functions doesn't change through upgrades. Otherwise this issue would have happened much earlier, before WASDK even became a thing when system types get updated with the Windows SDK.

I believe the core issue here is that WinUI 3 uses the same type names but changed the vtable layout and interface GUIDs. A similar issue would have happened when mixing C++/CX or WRL code using WinUI 2 with code using WinUI 3 (if WinUI 3 also supported those)

**Comment by Scottj1s** [comment]
Agreed - this is not specific to XAML, and warrants a fix.  The question is one of resourcing.

**Comment by kennykerr** [comment]
> I believe the core issue here is that WinUI 3 uses the same type names but changed the vtable layout and interface GUIDs. 

Seems like this would void your warranty. 😜

**Comment by dmachaj** [comment]
Note that WinAppSDK is __not__ ABI compatible across 1.x versions.  1.0 and 1.1 are __not__ ABI compatible, for example.  

(This is not how semantic versioning is supposed to work but I've been told that is the situation.  2.x versions, when they exist, __should__ follow proper semantic versioning).

**Comment by Scottj1s** [comment]
There were certainly ABI breaking changes between Project Reunion 0.8.* and WinAppSDK 1.*, and that is SemVer 2 compliant.  @dmachaj can you find an ABI break between WinAppSDK 1.0 and 1.2?  That would help justify the work.

**Comment by dmachaj** [comment]
Unfortunately I don't have any examples handy.  I am just repeating what I recall hearing previously.  I can ask around for an example.
